pkg/pyramid/file_tracker.go

```diff
@@ -0,0 +1,73 @@
+package pyramid
+
+import (
+	"sync"
+
+	"github.com/treeverse/lakefs/pkg/pyramid/params"
+)
+
+// fileTracker tracks file open requests in TierFS to avoid race conditions with cache rejection/eviction
+type fileTracker struct {
+	refMap map[string]*track
+	mu     sync.Mutex
+	delete deleteCallback
+}
+
+type deleteCallback func(path params.RelativePath)
+
+type track struct {
+	ref     int
+	deleted bool
+}
+
+func NewFileTracker(delete deleteCallback) *fileTracker {
+	return &fileTracker{
+		refMap: map[string]*track{},
+		mu:     sync.Mutex{},
+		delete: delete,
+	}
+}
+
+func (t *fileTracker) Open(path params.RelativePath) func() {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if val, ok := t.refMap[string(path)]; ok {
+		val.ref++
+	} else {
+		t.refMap[string(path)] = &track{
+			ref: 1,
+		}
+	}
+	return func() {
+		t.close(path)
+	}
+}
+
+func (t *fileTracker) close(path params.RelativePath) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if val, ok := t.refMap[string(path)]; ok {
+		val.ref--
+		if val.ref == 0 && val.deleted {
+			t.innerDelete(path)
+		}
```

Comment:
1. when val.ref == 0 - we need to delete it from the map
2. when we delete it from the map and val.deleted true - we need to invoke the callback